src/transform/src/analysis.rs

```diff
@@ -568,12 +577,81 @@ mod types {
                 .map(|child| &results[child])
                 .collect::<Vec<_>>();
 
-            if offsets.iter().all(|o| o.is_some()) {
-                let input_cols = offsets.into_iter().rev().map(|o| o.as_ref().unwrap());
-                let subtree_column_types = expr.try_col_with_input_cols(input_cols);
-                subtree_column_types.ok()
-            } else {
-                None
+            // For most expressions we'll apply `try_col_with_input_cols`, but for `Get` expressions
+            // we'll want to combine what we know (iteratively) with the stated `Get::typ`.
+            match expr {
+                MirRelationExpr::Get {
+                    id: mz_expr::Id::Local(i),
+                    typ,
+                    ..
+                } => {
+                    let mut result = typ.column_types.clone();
+                    if let Some(o) = depends.bindings().get(i) {
```

Comment:
They all start with `top()` from the `Lattice` implementor, which will be `None`. However (from the code) they get updated to `Some(_)` for each `Get` which should mean that no one (other than the `Get`) sees a `None`.

Reply:
Ah, sorry not sure I answered the right question. The `bindings()` should have each `i` we expect to see, but we are careful accessing them because in pessimistic mode we are less certain (at least, I worked from existing code rather than first principles on their behavior). However, once we get `o`, then we do expect `results[o]` to exist especially in optimistic mode (but not pessimistic).